src/capi/typeobject.cpp

```diff
@@ -623,6 +625,65 @@ static PyObject* slot_tp_iternext(PyObject* self) noexcept {
     return call_method(self, "next", &next_str, "()");
 }
 
+static PyObject* slot_tp_getattro(PyObject* self, PyObject* name) {
+    static PyObject* getattribute_str = NULL;
+    return call_method(self, "__getattribute__", &getattribute_str, "(O)", name);
+}
+
+static PyObject* call_attribute(PyObject* self, PyObject* attr, PyObject* name) {
+    PyObject* res, * descr = NULL;
+    descrgetfunc f = Py_TYPE(attr)->tp_descr_get;
+
+    if (f != NULL) {
+        descr = f(attr, self, (PyObject*)(Py_TYPE(self)));
+        if (descr == NULL)
+            return NULL;
+        else
+            attr = descr;
+    }
+    res = runtimeCall(attr, ArgPassSpec(1, 0, true, true), name, NULL, NULL, NULL, NULL);
+    Py_XDECREF(descr);
+    return res;
+}
+
+static PyObject* slot_tp_getattr_hook(PyObject* self, PyObject* name) {
+    PyObject* getattr, *getattribute, * res = NULL;
+    static PyObject* getattribute_str = NULL;
+    static PyObject* getattr_str = NULL;
+
+    /* speed hack: we could use lookup_maybe, but that would resolve the
+         method fully for each attribute lookup for classes with
+         __getattr__, even when the attribute is present. So we use
+         _PyType_Lookup and create the method only when needed, with
+         call_attribute. */
+    getattr = typeLookup(self->cls, _getattr_str, NULL);
+    if (getattr == NULL) {
+        /* No __getattr__ hook: use a simpler dispatcher */
+        self->cls->tp_getattro = slot_tp_getattro;
+        return slot_tp_getattro(self, name);
+    }
+    /* speed hack: we could use lookup_maybe, but that would resolve the
+         method fully for each attribute lookup for classes with
+         __getattr__, even when self has the default __getattribute__
+         method. So we use _PyType_Lookup and create the method only when
+         needed, with call_attribute. */
+    getattribute = typeLookup(self->cls, _getattribute_str, NULL);
+    try {
+        if (getattribute == NULL
+            || (Py_TYPE(getattribute) == wrapperdescr_cls
+                && ((BoxedWrapperDescriptor*)getattribute)->wrapped == (void*)PyObject_GenericGetAttr)) {
+            res = PyObject_GenericGetAttr(self, name);
+        } else {
+            res = call_attribute(self, getattribute, name);
+        }
+    } catch (ExcInfo e) {
+    }
+    if (res == NULL) {
+        res = runtimeCall(getattr, ArgPassSpec(2, 0, true, true), self, name, NULL, NULL, NULL);
```

Comment:
Similar note as above, runtimeCall throws Pyston exceptions, but slot_tp_getattr_hook is expected to throw CAPI exceptions, so we would need to convert here.  Adding a noexcept specifier is nice because it both annotates this and causes these kinds of mismatches to fail fast, since I think the compiler will convert any exceptions to std::terminate.